enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,518 @@
+memory _all_ the time. Instead, only _sometimes_ there are memory spikes
+within the virtual machine. Extrapolating this behavior to all virtual
+machines on a given node leads to the observation that _most often_ there
+is no memory pressure and a low memory utilization.
+Underutilized hardware resources are a cost factor to end-users.
+
+This proposal is about increasing the workload - virtual machine - density
+per node, in order to increase the memory utilization, in order to
+reduce the cost per virtual machine.
+
+> [!NOTE]
+> CPU resources are already over-committed in OpenShift today, thus the
+> limiting factor - usually - are memory resources.
+
+### User Stories
+
+* As an administrator, I want to be able to enable higher workload
+  density for OpenShift Virtualization clusters so that I can increase
+  the memory utilization.
+
+### Goals
+
+#### Timeline
+
+* GA higher workload density in OpenShift Virtualization in 2024
+
+#### Functionality
+
+* Fit more virtual machines onto a node once higher workload density
+  is enabled
+* Integrate well with [KSM] and [FPR]
+* **Technology Preview** - Enable higher density at all, limited
+  support for stressed clusters
+* **General Availability** - Improve handling of stressed clusters
+
+#### Usability
+
+* Provide a boolean in the OpenShift Console for enabling higher-density.
+  Do not require any additional workload or cluster level configuration
+  besides the initial provisioning.
+
+### Non-Goals
+
+* Complete life-cycling of the WASP Agent. We are not intending to write
+  an Operator for memory over commit for two reasons:
+  * [Kubernetes SWAP] is close, writing a fully fledged operator seems
+    to be no good use of resources
+  * To simplify the transition from WASP to [Kubernetes SWAP]
+
+## Proposal
+
+A higher workload density is achieved by combining two mechanisms
+
+1. Under-request memory resources for virtual machines by having the VM
+   pods (`launcher`) `requests.memory` being smaller than the VMs `guest.memory`
+   according to a configured ratio. This is owned by
+   [KubeVirt and present today](https://kubevirt.io/user-guide/operations/node_overcommit/).
+2. Compensate for the memory over-committment by using SWAP in order
+   to extend the virtual memory. This is owned by the platform and not
+   available today.
+
+#### Applicability
+
+We expect to mitigate the following situations
+
+* Underutilized guest memory (as described above)
+* Infrastructure related memory spikes in an over-committed environment
+* Considerable amount of "cold memory allocations" - thus allocations
+  which are rarely used
+
+#### Scope
+
+Memory over-committment, and as such swapping, will be limited to
+virtual machines running in the burstable QoS class.
+Virtual machines in the guaranteed QoS classes are not getting over
+committed due to alignment with upstream Kubernetes. Virtual machines
+will never be in the best-effort QoS because memory requests are
+always set.
```

Comment:
If I understand correctly, for GA, we also aim to make pods that are not associated with a VM use swap as well, right?

Reply:
Correct